src/middleware/dataset.ts

```diff
@@ -0,0 +1,25 @@
+import {
+  DefaultContext, DefaultState, Middleware, Next, Request, Response,
+} from 'koa';
+import { DataFactory, DatasetCore, DatasetCoreFactory } from 'rdf-js';
+import { DataFactoryContext } from './data-factory';
+
+export type WithDataset<T extends Request | Response> = T & {
+  dataset: DatasetCore;
+};
+
+export type ExtendedDataFactory = DataFactory & DatasetCoreFactory;
+
+export type DatasetContext<Context = DefaultContext> = DataFactoryContext<Context & {
+  request: WithDataset<Request>;
+  response: WithDataset<Response>;
+}, ExtendedDataFactory>;
+
+export default (): Middleware<DefaultState, DatasetContext> => (
+  async (context: DatasetContext, next: Next): Promise<void> => {
```

Comment:
doesn't this mean that `context` passed in here already is a `DatasetContext`, hence it has the `.request.dataset` and `.response.dataset` fields; if so, who has added them before these `Object.assign`?

Reply:
Think it was required when using `context.request = ...` with strict mode, but seems like it's ok with `Object.assign` so can remove.

Reply:
In fact, there's quite a few middleware where it's not quite right. Will push then extract.

Reply:
https://github.com/libero/article-store/pull/159